### PR TITLE
Use matcher.replace() to handle unicode characters with multiple code points

### PR DESCRIPTION
**Pull request by pcnfernando**
Fixes #[40315](https://github.com/pcnfernando/ballerina-lang/tree/issue-40315)

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: langlib/lang.regexp/src/main/ballerina/regexp.bal [hunk 1]
```diff
@@ -451,3 +435,8 @@ public isolated function fromString(string str) returns RegExp|error = @java:Met
     'class: "org.ballerinalang.langlib.regexp.FromString",
     name: "fromString"
 } external;
+
+isolated function replaceImpl(RegExp reExp, int startIndex, string str, string replacement) returns string = @java:Method {
```
Comment (SasinduDilshara) [inline langlib/lang.regexp/src/main/ballerina/regexp.bal:439]:
```suggestion
isolated function replaceImpl(RegExp regExp, int startIndex, string str, string replacement) returns string = @java:Method {
```

wdyt?

File: langlib/langlib-test/src/test/resources/test-src/regexp_test.bal [hunk 2]
```diff
@@ -579,6 +579,20 @@ function testReplace() {
     var regExpr5 = re `0+`;
     string result6 = regExpr5.replace(str6, replacementFunctionForReplace, 4);
     assertEquality("1001711", result6);
+
+    string str7 = "🔜 #RayoBarça 🔵🔴 https://t.co/iHDUx7EmFJ";
+    var regExpr6 = re `\p{S}`;
+    string result7 = regExpr6.replace(str7, replacementFunctionForReplace);
+    assertEquality("2 #RayoBarça 🔵🔴 https://t.co/iHDUx7EmFJ", result7);
```
Comment (SasinduDilshara) [inline langlib/langlib-test/src/test/resources/test-src/regexp_test.bal:586]: The symbol should be replace with value 1.
In the user point of view symbol is a one character. So the end index should be one in the replacement function.

In here, I think this gives 2 because symbol is represent by two code points, but Imo, when we give the results to users, we should consider it as one

File: langlib/langlib-test/src/test/resources/test-src/regexp_test.bal [hunk 3]
```diff
@@ -579,6 +579,20 @@ function testReplace() {
...
+
+    string result8 = regExpr6.replace(str7, replacementFunctionForReplace, 4);
+    assertEquality("🔜 #RayoBarça 16🔴 https://t.co/iHDUx7EmFJ", result8);
```
Comment (SasinduDilshara) [inline langlib/langlib-test/src/test/resources/test-src/regexp_test.bal:589]: The replacement value should be 14 in the user point of view. wdyt?

File: langlib/langlib-test/src/test/resources/test-src/regexp_test.bal [hunk 4]
```diff
@@ -624,12 +638,34 @@ function testReplaceAll() {
...
+    string str6 = "🔜 #RayoBarça 🔵🔴 https://t.co/iHDUx7EmFJ";
+    string result61 = regExpr5.replaceAll(str6, "");
+    assertEquality(" #RayoBarça  https://t.co/iHDUx7EmFJ", result61);
+    string result62 = regExpr5.replaceAll(str6, "", 12);
+    assertEquality("🔜 #RayoBarça  https://t.co/iHDUx7EmFJ", result62);
+
+    var regExpr6 = re `(\d{2})/(\d{2})/(\d{4})`;
+    string str7 = "04/05/2023, 05/05/2023";
+    string result71 = regExpr6.replaceAll(str7, updateDateFormat);
+    assertEquality("2023-05-04, 2023-05-05", result71);
```
Comment (SasinduDilshara) [inline langlib/langlib-test/src/test/resources/test-src/regexp_test.bal:652]: Shall we add a test for replace functions with `\p{sc=COMMON}` unicode property escape.

File: langlib/langlib-test/src/test/resources/test-src/regexp_test.bal [hunk 5]
```diff
@@ -1337,12 +1351,34 @@ function testReplaceAll() {
     var regExpr16_2 = re `AB|A|AK`;
     string result17_2 = regExpr16_2.replaceAll(str17, "Z");
     assertEquality("Z", result17_2);
+
+    var regExpr18 = re `\p{S}`;
+    string str18 = "🔜 #RayoBarça 🔵🔴 https://t.co/iHDUx7EmFJ";
+    string result18_1 = regExpr18.replaceAll(str18, "");
+    assertEquality(" #RayoBarça  https://t.co/iHDUx7EmFJ", result18_1);
+    string result18_2 = regExpr18.replaceAll(str18, "", 12);
+    assertEquality("🔜 #RayoBarça  https://t.co/iHDUx7EmFJ", result18_2);
```
Comment (SasinduDilshara) [inline langlib/langlib-test/src/test/resources/test-src/regexp_test.bal:1360]: Shall we add a one example that replace with a unicode character.

```
    string str18 = "🔜 #RayoBarça 🔵🔴 https://t.co/iHDUx7EmFJ";
    string result18_1 = regExpr18.replaceAll(str18, "🔴");
```

**Review by LakshanWeerasinghe — COMMENTED** [review]
Minor suggestion: Remove the url string in tests which points to an actual tweet.

**Review by LakshanWeerasinghe — APPROVED** [review]
Tested and LGTM!

File: langlib/langlib-test/src/test/resources/test-src/regexp_test.bal (hunk 5 above)
Comment (SasinduDilshara) [reply langlib/langlib-test/src/test/resources/test-src/regexp_test.bal:1360]: and something like

```
string str18 = "🔜 #RayoBarça 🔵🔴 https://t.co/iHDUx7EmFJ";
regexp:Span? result18_1 = re `🔵🔴`.find(str18);
```

Comment (SasinduDilshara) [reply langlib/langlib-test/src/test/resources/test-src/regexp_test.bal:1360]: I tested in locally and both are working fine, better to add an test to cover this.

File: langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/Matches.java [hunk 6]
```diff
@@ -36,8 +37,9 @@ public static BArray matchAt(BRegexpValue regExp, BString str, int startIndex) {
         checkIndexWithinRange(str, startIndex);
         Matcher matcher = RegexUtil.getMatcher(regExp, str);
         matcher.region(startIndex, str.length());
```
Comment (SasinduDilshara) [inline langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/Matches.java:39]: Don't we need to add 
```
int adjustedStartIndex = getSurrogateAdjustedStartIndex((int) startIndex, surrogates);
```
in here.

Comment (SasinduDilshara) [reply langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/Matches.java:39]: This is currently failing

```
import ballerina/lang.regexp;
import ballerina/io;

public function main() {
    string str18 = "🔜 #R";
    regexp:Span? result18_1 = re `#R`.matchAt(str18, 2);
    io:println((<regexp:Span> result18_1).substring(), " ", (<regexp:Span> result18_1).startIndex, " ", (<regexp:Span> result18_1).endIndex);
}   
```

Comment (pcnfernando) [reply langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/Matches.java:39]: Seems there are couple of other places as well eg: findAll. Will update

Comment (SasinduDilshara) [reply langlib/lang.regexp/src/main/java/org/ballerinalang/langlib/regexp/Matches.java]: Now it is working
tested for find Functions, match functions

**Review by SasinduDilshara — APPROVED** [review]

**Review by sanjana — APPROVED** [review]
